react/features/local-recording/controller/RecordingController.js

```diff
@@ -0,0 +1,493 @@
+                    this.onWarning(message);
+                }
+            }
+        }
+    }
+
+    /**
+     * Triggers the download of recorded data.
+     * Browser only.
+     *
+     * @param {number} sessionToken - The token of the session to download.
+     * @returns {void}
+     */
+    downloadRecordedData(sessionToken: number) {
+        if (this._adapters[sessionToken]) {
+            this._adapters[sessionToken].download();
+        } else {
+            logger.error(`Invalid session token for download ${sessionToken}`);
+        }
+    }
+
+    /**
+     * Switches the recording format.
+     *
+     * @param {string} newFormat - The new format.
+     * @returns {void}
+     */
+    switchFormat(newFormat: string) {
+        this._format = newFormat;
+        logger.log(`Recording format switched to ${newFormat}`);
+
+        // will be used next time
+    }
+
+    /**
+     * Returns the local recording stats.
+     *
+     * @returns {RecordingStats}
+     */
+    getLocalStats(): RecordingStats {
+        return {
+            currentSessionToken: this._currentSessionToken,
+            isRecording: this._state === ControllerState.RECORDING,
+            recordedBytes: 0,
+            recordedLength: 0
+        };
+    }
+
+    getParticipantsStats: () => *;
+
+    /**
+     * Returns the remote participants' local recording stats.
+     *
+     * @returns {*}
+     */
+    getParticipantsStats() {
+        const members
+            = this._conference.getParticipants()
+            .map(member => {
+                return {
+                    id: member.getId(),
+                    displayName: member.getDisplayName(),
+                    recordingStats:
+                        JSON.parse(member.getProperty(PROPERTY_STATS) || '{}'),
+                    isSelf: false
+                };
+            });
+
+        // transform into a dictionary,
+        // for consistent ordering
+        const result = {};
+
+        for (let i = 0; i < members.length; ++i) {
+            result[members[i].id] = members[i];
+        }
+        const localId = this._conference.myUserId();
+
+        result[localId] = {
+            id: localId,
+            displayName: i18next.t('localRecording.localUser'),
+            recordingStats: this.getLocalStats(),
+            isSelf: true
+        };
+
+        return result;
+    }
+
+    _updateStats: () => void;
+
+    /**
+     * Sends out updates about the local recording stats via XMPP.
+     *
+     * @private
+     * @returns {void}
+     */
+    _updateStats() {
+        if (this._conference) {
+            this._conference.setLocalParticipantProperty(PROPERTY_STATS,
+                JSON.stringify(this.getLocalStats()));
+        }
+    }
+
+    _onStartCommand: (*) => void;
+
+    /**
+     * Callback function for XMPP event.
+     *
+     * @private
+     * @param {*} value - The event args.
+     * @returns {void}
+     */
+    _onStartCommand(value) {
```

Comment:
What prevents a non-moderator from starting (or stopping) the recording?

Reply:
Nothing technically, it's just the use case that I have pictured: imagine recording an interview, normally I wouldn't want the guests to be messing around with the controls. However, there is indeed a loophole in here, because the "interviewer" might not necessarily be the one with the moderator rights.

Reply:
The question is whether or not the "local recording" rights should be aligned with moderator rights.